app/gui/src/dashboard/hooks/timeoutHooks.ts

```diff
@@ -122,3 +131,74 @@ export function useTimeout(params: Pick<UseTimeoutCallbackOptions, 'deps' | 'ms'
 
   return [isTimeout, restart] as const
 }
+
+/**
+ * Hook that returns a promise that resolves when the timeout expires.
+ */
+export function useTimeoutAPI(params: Pick<UseTimeoutCallbackOptions, 'ms'>) {
+  const { ms } = params
+
+  const timeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null)
+  const resolveRef = useRef<(() => void) | null>(null)
+  const rejectRef = useRef<((reason?: unknown) => void) | null>(null)
+
+  /**
+   * Restarts the timer.
+   */
+  const restartTimer = useEventCallback(async (signal?: AbortSignal) => {
+    stopTimer()
+    return startTimer(signal)
+  })
+
+  /**
+   * Starts the timer.
+   */
+  const startTimer = useEventCallback(async (signal?: AbortSignal) => {
+    stopTimer()
+
+    const abortHandler = () => {
+      rejectRef.current?.(new AbortError('Timeout aborted'))
+    }
+
+    const timeoutPromise = new Promise<void>((res, rej) => {
+      resolveRef.current = () => {
+        signal?.removeEventListener('abort', abortHandler)
+        res()
+      }
+      rejectRef.current = (...args: unknown[]) => {
+        signal?.removeEventListener('abort', abortHandler)
+        // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
+        rej(...args)
+      }
+    })
+
+    signal?.addEventListener('abort', abortHandler, { once: true })
+
+    timeoutRef.current = setTimeout(() => {
+      resolveRef.current?.()
+    }, ms)
+
+    return timeoutPromise
+  })
+
+  /**
+   * Stops the timer.
+   */
+  const stopTimer = useEventCallback(() => {
```

Comment:
optional: i would prefer defining this before `startTimer` just so readers know what `stopTimer` is before seeing the call